test/e2e/framework.go

```diff
@@ -1116,3 +1139,61 @@ func (data *TestData) GetGatewayInterfaceName(antreaNamespace string) (string, e
 	}
 	return antreaDefaultGW, nil
 }
+
+func (data *TestData) graceExitAntreaController(covDir string) error {
+	rc, stdout, _, err := provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl get pods -l=app=antrea,component=antrea-controller -n %s --no-headers=true", antreaNamespace))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when grace exit Antrea controller: get antrea-controller pod name, error: %v", err)
+	}
+	podName := strings.Split(stdout, " ")[0]
+	fmt.Printf("podName: %s\n", podName)
+	rc, stdout, _, err = provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl exec -i %s -n %s -- pgrep %s", podName, antreaNamespace, antreaControllerCovBinary))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when grace exit Antrea controller: pgrep %s, error: %v", antreaControllerCovBinary, err)
+	}
+	pid := stdout
+	rc, _, _, err = provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl exec -i %s -n %s -- kill -SIGINT %s", podName, antreaNamespace, pid))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when grace exit Antrea: send SIGINT signal, error: %v", err)
+	}
+	// a filepath-friendly timestamp format.
+	const timeFormat = "Jan02-15-04-05"
+	timeStamp := time.Now().Format(timeFormat)
+	rc, stdout, _, err = provider.RunCommandOnNode(masterNodeName(), fmt.Sprintf("kubectl cp %s/%s:%s %s/controller%s.out", antreaNamespace, podName, antreaControllerCovFile, covDir, timeStamp))
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when grace exit Antrea: copy %s from pod %s, error:%v", antreaControllerCovFile, podName, err)
+	}
+	return nil
+}
+
+func (data *TestData) graceExitAntreaAgent(covDir string, nodeName string) error {
```

Comment:
please add a detailed comment here explaining what this function is for

Reply:
ok, thanks.